File: solutions/3165. Maximum Sum of Subsequence With Non-adjacent Elements/3165.cpp
// Time complexity: O(n + q*sqrt(n))
// Space complexity: O(sqrt(n))
class Solution {
 public:
  int maximumSumSubsequence(vector<int>& nums,
                            const vector<vector<int>>& queries) {
    long long ans = 0;
    int partSize = max(8, static_cast<int>(sqrt(nums.size())));
    vector<array<int, 4>> parts;

    for (int i = 0; i < nums.size(); i += partSize)
      parts.push_back(calculateMaxNonAdjacentSum(nums, i, i + partSize));

    for (const auto& query : queries) {
      int partIndex = query[0] / partSize;
      nums[query[0]] = query[1];
      parts[partIndex] = calculateMaxNonAdjacentSum(nums, partIndex * partSize,
                                                    (partIndex + 1) * partSize);
      ans += combineParts(parts);
      ans %= kMOD;
    }

    return static_cast<int>(ans);
  }

 private:
  const int kMOD = 1'000'000'007;

  array<int, 4> calculateMaxNonAdjacentSum(vector<int>& nums, int start,
                                           int end) {
    int firstLast = 0;
    int firstSecondLast = 0;
    int secondLast = 0;
    int secondSecondLast = 0;

    for (int i = start; i < min(static_cast<int>(nums.size()), end); ++i) {
      firstLast =
          max(exchange(firstSecondLast, firstLast) + nums[i], firstLast);
      if (i > start)
        secondLast =
            max(exchange(secondSecondLast, secondLast) + nums[i], secondLast);
    }

    return {firstLast, firstSecondLast, secondLast, secondSecondLast};
  }

  int combineParts(vector<array<int, 4>>& parts) {
    int maxLast = 0;
    int maxSecondLast = 0;

    for (auto [firstLast, firstSecondLast, secondLast, secondSecondLast] :
         parts) {
      int temp = maxLast;
      maxLast = max(maxLast + secondLast, maxSecondLast + firstLast);
      maxSecondLast =
          max(temp + secondSecondLast, maxSecondLast + firstSecondLast);
    }

    return maxLast;
  }
};

File: solutions/3165. Maximum Sum of Subsequence With Non-adjacent Elements/3165.rs
// Time complexity: O(n + qsqrt(n))
// Space complexity: O(sqrt(n))
use std::cmp;
use std::mem;

impl Solution {
    pub fn maximum_sum_subsequence(mut nums: Vec<i32>, queries: Vec<Vec<i32>>) -> i32 {
        let mut ans = 0;
        let part_size = cmp::max(8, (nums.len() as f64).sqrt() as usize);
        let mut parts: Vec<[i32; 4]> = Vec::new();

        for i in (0..nums.len()).step_by(part_size) {
            parts.push(Self::calculate_max_non_adjacent_sum(
                &nums[i..cmp::min(i + part_size, nums.len())],
                0,
                part_size,
            ));
        }

        for query in queries {
            let part_index = query[0] as usize / part_size;
            nums[query[0] as usize] = query[1];
            parts[part_index] = Self::calculate_max_non_adjacent_sum(
                &nums[part_index * part_size..cmp::min((part_index + 1) * part_size, nums.len())],
                0,
                part_size,
            );
            ans += Self::combine_parts(&parts) as i64;
            ans %= Self::MOD;
        }

        ans as i32
    }

    const MOD: i64 = 1_000_000_007;

    fn calculate_max_non_adjacent_sum(nums: &[i32], start: usize, end: usize) -> [i32; 4] {
        let mut first_last = 0;
        let mut first_second_last = 0;
        let mut second_last = 0;
        let mut second_second_last = 0;

        for i in start..cmp::min(nums.len(), end) {
            first_last = cmp::max(
                mem::replace(&mut first_second_last, first_last) + nums[i],
                first_last,
            );
            if i > start {
                second_last = cmp::max(
                    mem::replace(&mut second_second_last, second_last) + nums[i],
                    second_last,
                );
            }
        }

        [
            first_last,
            first_second_last,
            second_last,
            second_second_last,
        ]
    }

    fn combine_parts(parts: &[[i32; 4]]) -> i32 {
        let mut max_last = 0;
        let mut max_second_last = 0;

        for &[first_last, first_second_last, second_last, second_second_last] in parts {
            let temp = max_last;
            max_last = cmp::max(max_last + second_last, max_second_last + first_last);
            max_second_last = cmp::max(
                temp + second_second_last,
                max_second_last + first_second_last,
            );
        }

        max_last
    }
}
